Build failure with PETSc master

When I try to build the newest libmesh GIT source together with PETSc master, I get the following error:

<pre>
src/solvers/tao_optimization_solver.C: In constructor ‘libMesh::TaoOptimizationSolver<T>::TaoOptimizationSolver(libMesh::OptimizationSystem&)’:
src/solvers/tao_optimization_solver.C:418:11: error: ‘TAO_CONVERGED_FATOL’ was not declared in this scope
   _reason(TAO_CONVERGED_FATOL/*==0*/) // Arbitrary initial value...
           ^
src/solvers/tao_optimization_solver.C: In instantiation of ‘void libMesh::TaoOptimizationSolver<T>::solve() [with T = double]’:
src/solvers/tao_optimization_solver.C:654:16:   required from here
src/solvers/tao_optimization_solver.C:502:26: error: too many arguments to function ‘PetscErrorCode TaoSetTolerances(Tao, PetscReal, PetscReal, PetscReal)’
   ierr = TaoSetTolerances(_tao,
                          ^
In file included from /home/tmac/usr/stow/petsc-real-dbg/include/petsc.h:27:0,
                 from ./include/libmesh/petsc_macro.h:93,
                 from ./include/libmesh/petsc_vector.h:31,
                 from src/solvers/tao_optimization_solver.C:28:
/home/tmac/usr/stow/petsc-real-dbg/include/petsctao.h:172:29: note: declared here
 PETSC_EXTERN PetscErrorCode TaoSetTolerances(Tao, PetscReal, PetscReal, PetscReal);
                             ^
Makefile:15865: recipe for target 'src/solvers/libmesh_dbg_la-tao_optimization_solver.lo' failed
make[2]: *** [src/solvers/libmesh_dbg_la-tao_optimization_solver.lo] Error 1
make[2]: Leaving directory '/home/tmac/src/upstream/libmesh'
</pre>

It is probably due to PETSc commit e52336c:

https://bitbucket.org/petsc/petsc/diff/include/petsctao.h?diff2=e52336cb213e&at=hzhang/matpackage-erroriffpe

Thanks for the report.  You can change it to TAO_CONVERGED_USER.

You'll also have to fix the TaoSetTolerances call in a fairly obvious way...

> On Nov 13, 2015, at 2:41 AM, torquil <notifications@github.com> wrote:
> 
> When I try to build the newest libmesh GIT source together with PETSc master, I get the following error:
> 
> src/solvers/tao_optimization_solver.C: In constructor ‘libMesh::TaoOptimizationSolver::TaoOptimizationSolver(libMesh::OptimizationSystem&)’:
> src/solvers/tao_optimization_solver.C:418:11: error: ‘TAO_CONVERGED_FATOL’ was not declared in this scope
>    _reason(TAO_CONVERGED_FATOL/*==0*/) // Arbitrary initial value...
>            ^
> src/solvers/tao_optimization_solver.C: In instantiation of ‘void libMesh::TaoOptimizationSolver::solve() [with T = double]’:
> src/solvers/tao_optimization_solver.C:654:16:   required from here
> src/solvers/tao_optimization_solver.C:502:26: error: too many arguments to function ‘PetscErrorCode TaoSetTolerances(Tao, PetscReal, PetscReal, PetscReal)’
>    ierr = TaoSetTolerances(_tao,
>                           ^
> In file included from /home/tmac/usr/stow/petsc-real-dbg/include/petsc.h:27:0,
>                  from ./include/libmesh/petsc_macro.h:93,
>                  from ./include/libmesh/petsc_vector.h:31,
>                  from src/solvers/tao_optimization_solver.C:28:
> /home/tmac/usr/stow/petsc-real-dbg/include/petsctao.h:172:29: note: declared here
>  PETSC_EXTERN PetscErrorCode TaoSetTolerances(Tao, PetscReal, PetscReal, PetscReal);
>                              ^
> Makefile:15865: recipe for target 'src/solvers/libmesh_dbg_la-tao_optimization_solver.lo' failed
> make[2]: *** [src/solvers/libmesh_dbg_la-tao_optimization_solver.lo] Error 1
> make[2]: Leaving directory '/home/tmac/src/upstream/libmesh'
> It is probably due to PETSc commit e52336c:
> 
> https://bitbucket.org/petsc/petsc/diff/include/petsctao.h?diff2=e52336cb213e&at=hzhang/matpackage-erroriffpe
> 
> —
> Reply to this email directly or view it on GitHub.
>

--- a/src/solvers/tao_optimization_solver.C
+++ b/src/solvers/tao_optimization_solver.C
@@ -500,8 +500,11 @@ void TaoOptimizationSolver<T>::solve ()
   // ||g(X)|| / ||g(X0)||                <= gttol
   // Command line equivalents: -tao_fatol, -tao_frtol, -tao_gatol, -tao_grtol, -tao_gttol
   ierr = TaoSetTolerances(_tao,
+#if PETSC_RELEASE_LESS_THAN(3,6,3)
+                          // Releases up to 3.6.2 had fatol and frtol, after that they were removed.
                           /*fatol=*/PETSC_DEFAULT,
                           /*frtol=*/this->objective_function_relative_tolerance,
+#endif
                           /*gatol=*/PETSC_DEFAULT,
                           /*grtol=*/PETSC_DEFAULT,
                           /*gttol=*/PETSC_DEFAULT);

--- a/src/solvers/tao_optimization_solver.C
+++ b/src/solvers/tao_optimization_solver.C
@@ -415,7 +415,7 @@ template <typename T>
 TaoOptimizationSolver<T>::TaoOptimizationSolver (OptimizationSystem& system_in)
   :
   OptimizationSolver<T>(system_in),
-  _reason(TAO_CONVERGED_FATOL/*==0*/) // Arbitrary initial value...
+  _reason(TAO_CONVERGED_USER) // Arbitrary initial value...
 {
 }
 

